components/remerge/src/storage/db.rs

```diff
@@ -57,4 +63,96 @@ impl RemergeDb {
         tx.commit()?;
         Ok(RemergeDb { db, info })
     }
+
+    pub fn exists(&self, id: &str) -> Result<bool> {
+        Ok(self.db.query_row_named(
+            "SELECT EXISTS(
+                 SELECT 1 FROM rec_local
+                 WHERE guid = :guid AND is_deleted = 0
+                 UNION ALL
+                 SELECT 1 FROM rec_mirror
+                 WHERE guid = :guid AND is_overridden IS NOT 1
+             )",
+            rusqlite::named_params! { ":guid": id },
+            |row| row.get(0),
+        )?)
+    }
+
+    pub fn create(&self, record_info: JsonValue) -> Result<sync_guid::Guid> {
+        let mut id = Guid::random();
+        let mut to_insert = serde_json::Map::default();
+        let record_obj = record_info
+            .as_object()
+            .ok_or_else(|| InvalidRecord::NotJsonObject)?;
+        for field in &self.info.local.fields {
+            let native_field = &self
+                .info
+                .native
+                .fields
+                .iter()
+                .find(|f| f.name == field.name);
+            let local_name = native_field
+                .map(|n| n.local_name.as_str())
+                .unwrap_or_else(|| field.name.as_str());
+            let is_guid = crate::schema::FieldKind::OwnGuid == field.ty.kind();
+            if let Some(v) = record_obj.get(local_name) {
+                let fixed = field.validate(v.clone())?;
+                if is_guid {
+                    if let JsonValue::String(s) = &fixed {
+                        id = Guid::from(s.as_str());
+                    } else {
+                        unreachable!("Field::validate checks this");
+                    }
+                }
+                to_insert.insert(field.name.clone(), fixed);
+            } else if let Some(def) = field.ty.get_default() {
+                to_insert.insert(field.name.clone(), def);
+            } else if is_guid {
+                to_insert.insert(field.name.clone(), id.to_string().into());
+            } else if field.required {
+                throw!(InvalidRecord::MissingRequiredField(local_name.to_owned()));
+            }
+        }
+        let tx = self.db.unchecked_transaction()?;
+        // TODO: Search DB for dupes based on the value of the fields listed in dedupe_on.
+        let id_exists = self.exists(id.as_ref())?;
+        if id_exists {
+            throw!(InvalidRecord::IdNotUnique);
+        }
+        let ctr = super::meta::get::<i64>(&self.db, super::meta::CHANGE_COUNTER)? + 1;
```

Comment:
I would have expected the change counter to be per-record/per-vector-clock. I don't think there's a problem per se with it being global, except maybe a long-shot possibility of it wrapping?

Reply:
There's a comment in the third commit explaining that there's basically no realistic chance of it overflowing. Out of curiosity, is there an advantage to it being global?